otter/convergence/steps.py

```diff
@@ -70,6 +70,46 @@ def set_server_name(server_config_args, name_suffix):
     return set_in(server_config_args, ('server', 'name'), name)
 
 
+def _failure_reporter(*terminal_err_types):
+    """
+    Return a callable that takes an error tuple which interprets the error
+    tuple.
+
+    If the error is an APIError with status code 4xx, or one of the provided
+    ``terminal_err_types``, then the callable returns a tuple of::
+
+        (StepResult.FAILURE, [ErrorReason.Exception(exc_tuple)])
+
+    else it returns a tuple of::
+
+        (StepResult.RETRY, [ErrorReason.Exception(exc_tuple)])
+    """
+    def reporter(exc_tuple):
+        err_type, error, traceback = exc_tuple
+
+        terminal_error = (
+            err_type in terminal_err_types or
+            (err_type == APIError and 400 <= error.code < 500)
+        )
+
+        if terminal_error:
+            return StepResult.FAILURE, [ErrorReason.Exception(exc_tuple)]
+        return StepResult.RETRY, [ErrorReason.Exception(exc_tuple)]
+
+    return reporter
+
+
+def _success_reporter(success_reason):
```

Comment:
Ah ok, I was interpreting it as the cloud client succeeded, and how to report that success, same with `_failure_reporter`, since it also can return RETRY as well as FAILURE.